packages/block-library/src/block/edit.js

```diff
@@ -6,207 +6,188 @@ import { partial } from 'lodash';
 /**
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
-import { Placeholder, Spinner, Disabled } from '@wordpress/components';
-import { withSelect, withDispatch } from '@wordpress/data';
-import { __ } from '@wordpress/i18n';
 import {
 	BlockEditorProvider,
 	BlockList,
 	WritingFlow,
 } from '@wordpress/block-editor';
-import { compose } from '@wordpress/compose';
 import { parse, serialize } from '@wordpress/blocks';
+import { Placeholder, Spinner, Disabled } from '@wordpress/components';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
 import ReusableBlockEditPanel from './edit-panel';
 
-class ReusableBlockEdit extends Component {
-	constructor( { reusableBlock } ) {
-		super( ...arguments );
-
-		this.startEditing = this.startEditing.bind( this );
-		this.stopEditing = this.stopEditing.bind( this );
-		this.setBlocks = this.setBlocks.bind( this );
-		this.setTitle = this.setTitle.bind( this );
-		this.save = this.save.bind( this );
-
-		if ( reusableBlock ) {
-			// Start in edit mode when we're working with a newly created reusable block
-			this.state = {
-				isEditing: reusableBlock.isTemporary,
-				title: reusableBlock.title,
-				blocks: parse( reusableBlock.content ),
-			};
-		} else {
-			// Start in preview mode when we're working with an existing reusable block
-			this.state = {
-				isEditing: false,
-				title: null,
-				blocks: [],
+const EMPTY_ARRAY = [];
+
+export default function ReusableBlockEdit( { attributes, isSelected } ) {
+	const { ref } = attributes;
+
+	const {
+		theReusableBlock,
+		isFetching,
+		isSaving,
+		isTemporary,
+		blocks,
+		canUpdateBlock,
+		settings,
+		title,
+	} = useSelect(
+		/** @typedef {import('@wordpress/blocks').BlockInstance} BlockInstance */
+
+		/**
+		 * @typedef WPReusableBlockEditSelectedData
```

Comment:
I guess there's value in having it located near where it's used, but I'd personally prefer most custom type definitions to be consolidated toward the top of a file, so as to be a consistent reference point. It also helps avoid the issue you've run into with line length on the property descriptions.

Reply:
Also, this doesn't feel like something we should have to create a type for. If the selector return value is properly typed, then TypeScript should be able to infer the return value for us, which would go a long way to avoid human error and duplicate typing.

I kinda feel all this typing should be removed. I don't think we can or should expect developers to add types like this in every usage of `useSelect`, and if we do, it's a failing on the part of the framework tooling.

Reply:
Currently, the types aren't inferred because the functions returned from the `select` calls don't have type definitions. Now that you point it out, the real solution would be to add types to those where they're defined. I'll remove the type definition from this PR.